--- a/opencv/cmake/OpenCVCompilerOptions.cmake	2016-01-14 10:30:06.617458710 +0200
+++ b/opencv/cmake/OpenCVCompilerOptions.cmake	2016-01-14 12:20:50.449304819 +0200
@@ -102,7 +102,7 @@
     add_extra_compiler_option(-Werror)
   endif()
 
-  if(X86 AND NOT MINGW64 AND NOT X86_64 AND NOT APPLE)
+  if(X86 AND NOT MINGW64 AND NOT X86_64 AND NOT APPLE AND NOT EMSCRIPTEN)
     add_extra_compiler_option(-march=i686)
   endif()
 
