--- ./clang//fastcomp/src/lib/Transforms/Scalar/GVN.cpp	2016-03-10 11:56:33.968255293 +0200
+++ ./clang//fastcomp/src/lib/Transforms/Scalar/GVN.cpp	2016-03-10 11:55:06.252253874 +0200
@@ -1893,6 +1893,16 @@
 
   // ... to a pointer that has been loaded from before...
   MemDepResult Dep = MD->getDependency(L);
+
+  if (Dep.getInst()) {
+    if (Dep.getInst()->getType()->isVectorTy() ^ L->getType()->isVectorTy())
+      return false;
+    if (StoreInst *DepSI = dyn_cast<StoreInst>(Dep.getInst())) {
+      if (DepSI->getOperand(0)->getType()->isVectorTy() ^ L->getType()->isVectorTy())
+        return false;
+    }
+  }
+
   const DataLayout &DL = L->getModule()->getDataLayout();
 
   // If we have a clobber and target data is around, see if this is a clobber
